azure-blob-storage: utilize streaming capabilities - #1614

denisneuling opened the pull request:
#### What is the purpose of this change?

Replacing the multipart upload from the azureblob backend by the streaming capabilities provided via [azure-storage-blob-go](https://github.com/Azure/azure-storage-blob-go). 
This change will eliminate the need to spool a file going to be uploaded; means disk consumption will be reduced when transferring files into an azure blob container and thus making life easier when e.g. backing up huge files where disk space is critical. 

#### Was the change discussed in an issue or in the forum before?

#1614

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

sandeepkru commented on backend/azureblob/azureblob.go:
```diff
@@ -1403,26 +1229,12 @@ func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, op
-			blockBlobURL := blob.ToBlockBlobURL()
-			_, err = azblob.UploadStreamToBlockBlob(ctx, in, blockBlobURL, putBlobOptions)
-		}
+		// Write a small blob in one transaction
```
Nit: - Update this comment?

sandeepkru commented:
Thanks for sending out this PR, I've couple of questions and minor comments.
1. How MD5 comparisons work here for large blobs? for example if blob is 5TB? Are we reading entire blob here - https://github.com/rclone/rclone/blob/master/backend/azureblob/azureblob.go#L1391 and streaming for upload again?
2. Who handles block list with `Put` of large blocks? Is azure go SDK automatically creating block list? If yes, what is the block size, and how do they handle partial uploads? or did azure storage eliminate block list with streaming upload of large blobs?(I didn't check their documentation recently)
3. Can you please revisit these comments https://github.com/rclone/rclone/blob/master/backend/azureblob/azureblob.go#L1387-L1391

Can you please check and address failing tests?

ncw commented:
The CI tests look like they are passing now (except for one unrelated race we haven't seen before).

@sandeepkru are you happy with this now?

@denisneuling did you run the integration tests?

I ran them locally and they all passed :-)


```
go install ./...
test_all -backends azureblob
```

![image](https://user-images.githubusercontent.com/536803/65039784-f273e680-d94a-11e9-8123-6721e2fbda22.png)

(These weren't done on the master branch - that is a bug which needs fixing!)

denisneuling commented:
First at all many thanks for the quick review and sorry for the belated reply from my side.

I've fixed the tests which were mainly failing due to an overseen setter function in the azureblob_test.go file and due to coding style issues - these issues have been resolved and also all (integration) tests ran fine on my local machine. I am bit wondering why they're failing on travis-ci - unfortunately I havn't had spare time to look into that issue yet. 
Also couldn't investigate about the block list handling @sandeepkru was mentioning, but was planning to investigate about that by end of this week. 
What i have found regarding the MD5 checksums so far by having a quick look into the azure-storage-blob-go SDK was that checksums are not going to be used when using its streaming capabilities - so I also removed the section where the MD5 have been calculated.

sandeepkru commented:
MD5 aspect is always debatable, see here - https://github.com/Azure/azure-storage-blob-go/issues/56. However, azure has this MD5 support for small blobs at entire blob level or for large blobs at block level(there is no TreeHash like checksum (like S3) at entire blob level though for large blobs).

- Is it ok for rclone users not to compute MD5 on all uploads? 

I would like to try quick testing on this locally sometime later today and verify how blocklist managed, hope it can wait for one more day.

ncw commented:
> What i have found regarding the MD5 checksums so far by having a quick look into the azure-storage-blob-go SDK was that checksums are not going to be used when using its streaming capabilities - so I also removed the section where the MD5 have been calculated

Will the blobs have MD5s on when uploaded?

ncw commented:
Where are we with PR?  @sandeepkru are we ready to merge?

@denisneuling apologies for the delay

ivandeex commented:
@sandeepkru any news?

ivandeex commented:
@sandeepkru

ivandeex commented:
@denisneuling 
Please accept https://github.com/denisneuling/rclone/pull/1 to reopen this PR.

ivandeex commented:
@ncw 
This PR is rebased. Please review.
Note that @sandeepkru has no activity on github since June 2019 (https://github.com/sandeepkru?tab=overview&from=2019-08-01&to=2020-12-30)
I suggest putting this PR in a branch, release a beta on beta.rclone.org and call for beta testers.

ivandeex commented:
Rebased again on latest master to get rid of internal merge.

ivandeex commented:
@denisneuling @ncw
Please pay attention to the following chunk:
https://github.com/rclone/rclone/pull/3533/files#diff-52c116af5a1db57e60138a79e7a227a6eb7ed8efa9b6842b02b58b00540a5aa1R1290
When @denisneuling submitted this PR he completely removed it (see https://github.com/ivalearn/rclone-denisneuling/commit/6db4d04b303ff16e33269fda2493522af76b5dfb#diff-52c116af5a1db57e60138a79e7a227a6eb7ed8efa9b6842b02b58b00540a5aa1L1386)
On rclone master that code has become conditional, so I'm not sure whether removing it is still correct. It's probably related to MD5 https://github.com/rclone/rclone/pull/3533#issuecomment-532302582
Please review it.

ncw commented:
I reviewed the code - I think this is a great improvement pushing the heavy lifting to the azureblob SDK

> Please pay attention to the following chunk:
> https://github.com/rclone/rclone/pull/3533/files#diff-52c116af5a1db57e60138a79e7a227a6eb7ed8efa9b6842b02b58b00540a5aa1R1290
> When @denisneuling submitted this PR he completely removed it (see [ivalearn@6db4d04#diff-52c116af5a1db57e60138a79e7a227a6eb7ed8efa9b6842b02b58b00540a5aa1L1386](https://github.com/ivalearn/rclone-denisneuling/commit/6db4d04b303ff16e33269fda2493522af76b5dfb#diff-52c116af5a1db57e60138a79e7a227a6eb7ed8efa9b6842b02b58b00540a5aa1L1386))
> On rclone master that code has become conditional, so I'm not sure whether removing it is still correct. It's probably related to MD5 [#3533 (comment)](https://github.com/rclone/rclone/pull/3533#issuecomment-532302582)
> Please review it.

We need that chunk of code - if we don't have it then we don't set MD5SUMs on uploaded objects.

The comment should say something like

```go
	// Compute the Content-MD5 of the file. As we stream all uploads it
	// will be set in PutBlockList API call using the 'x-ms-blob-content-md5' header
	if !o.fs.opt.DisableCheckSum {
```

If you fix that up and the conflict I'll merge it - thank you @ivandeex :-)

ivandeex commented:
> I reviewed the code - I think this is a great improvement pushing the heavy lifting to the azureblob SDK
> We need that chunk of code - if we don't have it then we don't set MD5SUMs on uploaded objects.
> If you fix that up and the conflict I'll merge it

@ncw Done

ivandeex commented:
@ncw :cry: :pray: :kneeling_woman: :sob:

ncw commented:
Thanks @ivandeex 

I gave this a test and it is looking good.

Thank you @denisneuling for the patch and apologies it has taken so long to merge.
